### PR TITLE
[ADD][8.0] website_product_show_uom

### PR DESCRIPTION
**Pull request by andhit-r**
# Show UoM

This module brings the users the possibility to:
1. Show/hide UoM on product list page
2. Show/hide UoM on all product pages
3. Show/hide UoM on checkout page
# Usage

**Show UoM on product list page**
1. Go to website
2. Open shop menu
3. Click **Customize** menu
4. Toggle **Show UoM** option to show/hide UoM

UoM will show after product price

**Show UoM on each product page(s)**
1. Go to website
2. Open shop menu
3. Open one product page
4. Click **Customize** menu
5. Toggle **Show UoM** option to show/hide UoM

UoM will show after product price

**Show UoM on checkout page**
1. Go to website
2. Open shop menu
3. Open one product page
4. Click **Add to chart** button. This will open checkout page
5. Click **Customize** menu
6. Toggle **Show UoM** option to show/hide UoM

New column for UoM will show after quantity on Shopping Chart table

### CODE REVIEW

File: website_product_show_uom/README.rst [hunk 1]
```diff
@@ -0,0 +1,84 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License
```
Comment (pedrobaeza) [inline website_product_show_uom/README.rst]: Please include link like in https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst

File: website_product_show_uom/README.rst [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+Show UoM
```
Comment (pedrobaeza) [inline website_product_show_uom/README.rst]: Put the upper line

**Comment by oscarolar** [comment]
Hello @andhit-r as I expected your app works only logged in, since specific permissions are required to UoMs, can you check it out without being logged in?, here is a screenshot:
screen

I also suggest that you make some PhantomJS tests with 4 users: Admin, Demo, Portal and Public(no user logged).

**Comment by cvinh** [comment]
Hello @andhit-r 
I think 2 features are missing :
- "Show UOM" Option should be shown in payment page
- UOM should be changed by the customer
  Best regards

**Comment by andhit-r** [comment]
@oscarolar thanks for your review. Access rule for public and portal user have been added. I'm not to familiar with PhantomJS test. Would you point me to it's documentation.

@invitu thanks for your review. Show UoM option has been added to payment page. You could check it on runbot. The purpose of this module only to show UoM, not to give the ability to change order line UoM. So i will not add that functionality to this module. Sorry.

**Comment by cvinh** [comment]
Dear andhit-r,
In some case, it would be very helpful to let the customer choose the unit : for example, if you have a product available in boxes or cases, the customer could choose between the 2 packages.
If you don't want the customer to be able to choose, you could lock it with readonly attribute or security.
Best regards

**Comment by andhit-r** [comment]
@invitu it might be helpful. But as i explained in my previous comment, i will not add that functionality in this module since that is not the purpose of this module.

**Comment by oscarolar** [comment]
I made a video but is in spanish I havent had time to make it in english:
https://www.youtube.com/watch?v=0JXulEZWtBc

**Step 1:** Include the PhantomJS Tour: 
https://github.com/odoo/odoo/blob/8.0/addons/website_sale/views/templates.xml#L14
**Step 2** declare the test modes:
https://github.com/odoo/odoo/blob/8.0/addons/website_sale/tests/test_sale_process.py#L3
**Step 3** write the actual test:
https://github.com/odoo/odoo/blob/8.0/addons/website_sale/static/src/js/website_sale_tour_buy.js

I think of it as a simple test, something like: enable the feature via demo data and then check with all users that at least one product has the uom after the price.

File: website_product_show_uom/README.rst [hunk 3]
```diff
@@ -0,0 +1,101 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+=========
+Show UoM
```
Comment (elicoidal) [inline website_product_show_uom/README.rst]: Display Unit of Measure in e-commerce

File: website_product_show_uom/README.rst [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+=========
+This module brings the users the possibility to:
+
+1. Show/hide UoM on product list page
```
Comment (elicoidal) [inline website_product_show_uom/README.rst]: Show/hide Unit of Measure (UoM) on product list page

File: website_product_show_uom/README.rst [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+=====
+
+**Show UoM on product list page**
+
+1. Go to website
+2. Open shop menu
+3. Click **Customize** menu
+4. Toggle **Show UoM** option to show/hide UoM
+
+UoM will show after product price
```
Comment (elicoidal) [inline website_product_show_uom/README.rst]: period

File: website_product_show_uom/README.rst [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+**Show UoM on each product page(s)**
+
+1. Go to website
+2. Open shop menu
+3. Open one product page
+4. Click **Customize** menu
+5. Toggle **Show UoM** option to show/hide UoM
+
+UoM will show after product price
```
Comment (elicoidal) [inline website_product_show_uom/README.rst]: UoM will be displayed after the product price.

File: website_product_show_uom/README.rst (hunk 4 above)
Comment (elicoidal) [inline website_product_show_uom/README.rst]: Better "Display" than Show (several)

File: website_product_show_uom/README.rst [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+**Show UoM on checkout page**
+
+1. Go to website
+2. Open shop menu
+3. Open one product page
+4. Click **Add to chart** button. This will open checkout page
+5. Click **Customize** menu
+6. Toggle **Show UoM** option to show/hide UoM
+
+New column for UoM will show after quantity on Shopping Chart table
```
Comment (elicoidal) [inline website_product_show_uom/README.rst]: A new column for UoM will be displayed after the quantity on Shopping Chart table.

File: website_product_show_uom/README.rst [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+1. Go to website
+2. Open shop menu
+3. Open one product page
+4. Click **Add to chart** button. This will open checkout page
+5. Click **Process Checkout** button. This will open billing information page
+6. Click **Confirm** button. This will open validate order page
+7. Click **Customize** menu
+8. Toggle **Show UoM** option to show/hide UoM
+
+New column for UoM will show after quantity on Shopping Chart table
```
Comment (elicoidal) [inline website_product_show_uom/README.rst]: A new column for UoM will be displayed after the quantity on Shopping Chart table.

File: website_product_show_uom/README.rst [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+Know issues / Roadmap
+=====================
+* Show UoM on product page(s) configuration valid for all products. No individual configuration for each product.
```
Comment (elicoidal) [inline website_product_show_uom/README.rst]: - Display UoM on product page(s) configuration valid for all products (No individual configuration for each product).

File: website_product_show_uom/README.rst [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+* This module does not give the ability to change order line UoM
```
Comment (elicoidal) [inline website_product_show_uom/README.rst]: period

File: website_product_show_uom/README.rst [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/113/51
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/e-commerce/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/e-commerce/issues/new?body=module:%20website_product_share%0Aversion:%208.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
Comment (elicoidal) [inline website_product_show_uom/README.rst]: update

File: website_product_show_uom/__openerp__.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+# © 2016 Andhitia Rama <andhitia.r@gmail.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+{
+    "name": "Show Product UoM",
```
Comment (elicoidal) [inline website_product_show_uom/__openerp__.py]: Display Product UoM in e-commerce website

**Comment by andhit-r** [comment]
@oscarolar i have done simple test like you said. Would you review it again.

@elicoidal revision done.

Thanks in advance

**Comment by oscarolar** [comment]
@andhit-r I dont see any test

**Comment by andhit-r** [comment]
@oscarolar what i mean that i have tested manually, not created a test file. I have studied your pointer regarding PhantomJS test and still trying to understand it. Could someone point me to other documentation regarding this PhantomJS test. 

I don't see any PhantomJS test on OCA/website modules. Is it mandatory?

**Comment by oscarolar** [comment]
Its not mandatory but as you can see this was a "False Green" since there are no tests, with the PhantomJS test you ensure that this is a "Real Green", anyway I made the test for you and a PR to your repo https://github.com/andhit-r/e-commerce/issues/1

Best regards.

**Comment by andhit-r** [comment]
@oscarolar Thank you for PhantomJS example. I will add more test for other page.

**Comment by andhit-r** [comment]
@oscarolar i added more phantomJS test. Would you kindly review it?

@elicoidal  would you kindly put **need review** label on this PR

Thanks is advance

File: website_product_show_uom/tests/test_website_uom_logout.py [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+import openerp.tests
+
+
+@openerp.tests.common.at_install(False)
+@openerp.tests.common.post_install(True)
+class TestUi(openerp.tests.HttpCase):
+    def test_01_public_uom(self):
+        self.phantom_js(
+            "/", "openerp.Tour.run('website_test_uom_logout', 'test')",
+            "openerp.Tour.tours.website_test_uom_logout")
```
Comment (oscarolar) [inline website_product_show_uom/tests/test_website_uom_logout.py]: Thanks a lot @andhit-r ,
I don't understand why you are creating another test for a not logged in user, the previous test `website_test_uom_login` can be used with no login is as simple as adding an extra test method like this:

``` python
    def test_03_public_uom(self):
        self.phantom_js(
            "/", "openerp.Tour.run('website_test_uom_login', 'test')",
            "openerp.Tour.tours.website_test_uom_login")
```

just removing the `login` param as you did on this test should work.

Is there something I'm not seeing?

**Comment by andhit-r** [comment]
@oscarolar tried that. It always make the test failed. It failed when public user click **checkout button**. I guess it has something to do with **website_require_login** or **website_suggest_create_account**. Both modules alter website sales flow and both modules tested.

So i decided to split test into 2 tests. One for admin and demo (it tests from start flow untill payment screen). One for public (it test from start till shopping cart screen to avoid interaction with non-standard flow)

**Comment by oscarolar** [comment]
:+1:

**Comment by andhit-r** [comment]
Hi @OCA/e-commerce-maintainers 

Would you kindly review this PR. Thanks in advance.

**Comment by rafaelbn** [comment]
Tested in runbot :+1:

**Comment by mikevhe18** [comment]
Tested in runbot :+1:

**Comment by andhit-r** [comment]
Hi @OCA/e-commerce-maintainers this PR already has 3 upvote and green checks. Is there anything i still have to do? If not then this PR is ready to be merged. Many thanks

**Comment by andhit-r** [comment]
Thanks @pedrobaeza . Thank you all reviewers

**Pull request merged**